Draft-Accepted/RFCXXXX-Command-Shell-Integration.md

```diff
@@ -0,0 +1,287 @@
+
+If the tool calls other executables, it is responsible for removing or propagating the environment variable which
+can affect the behavior of child processes.
+
+JSON output should include a `$typeNames` member which is an array of strings that indicate the type hierarchy of the object.
+PowerShell would use this as the same as the `TypeNames` member of a PSObject for formatting.
+Future enhancement would support a JSON defined formatting schema that could be used by other shells.
+
+### Tab Completion Registration
+
+Within the same manifest, a tool can indicate it supports tab completion.
+
+This can either be an executable with arguments.
+
+Arguments can be built-in variables to provide context for the tab completion.
+
+Currently supported variables:
+
+- `{commandLine}` - The full command line that the user has typed so far.
+- `{cursorPosition}` - The current cursor position within the command line, starting at 0 index.
+
+```json
+{
+    "$schema": "https://schemas.microsoft.com/commandshell/2023/05/01/command.shellintegration.json",
+    "executable": "az",
+    "version": "1.0.0",
+    "tabCompletion": {
+        "command": {
+            "executable": "az",
+            "arguments": [
+                "completion",
+                "powershell",
+                "--commandLine",
+                "{commandLine}",
+                "--cursorPosition",
+                "{cursorPosition}"
+            ]
+        }
+    }
+}
+```
+
+or a shell specific script:
+
+```json
+{
+    "$schema": "https://schemas.microsoft.com/commandshell/2023/05/01/command.shellintegration.json",
+    "executable": "az",
+    "version": "1.0.0",
+    "tabCompletion": {
+        "script": {
+            "powershell": "az-completion.ps1",
+            "bash": "az-completion.sh",
+            "zsh": "az-completion.zsh"
+        }
+    }
+}
+```
+
+Multiple shells can be specified and up to the shell to determine which one to use.
+Both a `command` and `script` can be specified, but only one of each and also up to the shell to determine which one to use.
+
+### Help Registration
+
+The same manifest may define a path to the help content:
+
+```json
+{
+    "$schema": "https://schemas.microsoft.com/commandshell/2023/05/01/command.shellintegration.json",
+    "executable": "az",
+    "version": "1.0.0",
+    "help": {
+        "tags": [
+            "azure",
+            "az"
+        ],
+        "path": "./help/az-help.md"
+    }
+}
+```
+
+Optional `tags` can be specified to help with searching for help content.
+
+### Predictors and Feedback Providers Registration
+
+Specifically for PowerShell, a command may also want to optionally register a Predictor or Feedback Provider.
+
+```json
+{
+    "$schema": "https://schemas.microsoft.com/commandshell/2023/05/01/command.shellintegration.json",
+    "executable": "az",
+    "version": "1.0.0",
+    "predictor": {
```

Comment:
Same here, we need to also consider predictors and feedback providers that are shipped in PowerShell module, and allow them to be registered to the shell without requiring updating the user's profile. It would be important to have one mechanism that works for both native tool and PowerShell module scenarios.

Reply:
As noted above, for modules it makes sense to have something in the Module manifest under PrivateData but is outside the scope of this spec.  Adding more details to support MSIX scenario.

Reply:
> but is outside the scope of this spec

I think this spec should consider that as well, see comment https://github.com/PowerShell/PowerShell-RFC/pull/348#discussion_r1181704869

Reply:
added